nexus/types/src/inventory/display.rs

```diff
@@ -0,0 +1,941 @@
+};
+use strum::IntoEnumIterator;
+use tabled::Tabled;
+use tufaceous_artifact::ArtifactHash;
+use uuid::Uuid;
+
+use crate::inventory::{
+    CabooseWhich, Collection, Dataset, PhysicalDisk, RotPageWhich, Zpool,
+};
+
+/// Code to display inventory collections.
+pub struct CollectionDisplay<'a> {
+    collection: &'a Collection,
+    include_sps: CollectionDisplayIncludeSps,
+    include_sleds: bool,
+    include_orphaned_datasets: bool,
+    include_clickhouse_keeper_membership: bool,
+    long_string_formatter: LongStringFormatter,
+}
+
+impl<'a> CollectionDisplay<'a> {
+    pub(super) fn new(collection: &'a Collection) -> Self {
+        Self {
+            collection,
+            // Display all items by default.
+            include_sps: CollectionDisplayIncludeSps::None,
+            include_sleds: true,
+            include_orphaned_datasets: true,
+            include_clickhouse_keeper_membership: true,
+            long_string_formatter: LongStringFormatter::new(),
+        }
+    }
+
+    /// Control display of SP information (defaults to
+    /// [`CollectionDisplayIncludeSps::All`]).
+    pub fn include_sps(
+        &mut self,
+        include_sps: CollectionDisplayIncludeSps,
+    ) -> &mut Self {
+        self.include_sps = include_sps;
+        self
+    }
+
+    /// Control display of sled information (defaults to true).
+    pub fn include_sleds(&mut self, include_sleds: bool) -> &mut Self {
+        self.include_sleds = include_sleds;
+        self
+    }
+
+    /// Control display of orphaned datasets (defaults to true).
+    pub fn include_orphaned_datasets(
+        &mut self,
+        include_orphaned_datasets: bool,
+    ) -> &mut Self {
+        self.include_orphaned_datasets = include_orphaned_datasets;
+        self
+    }
+
+    /// Control display of ClickHouse keeper membership information (defaults to
+    /// true).
+    pub fn include_clickhouse_keeper_membership(
+        &mut self,
+        include_clickhouse_keeper_membership: bool,
+    ) -> &mut Self {
+        self.include_clickhouse_keeper_membership =
+            include_clickhouse_keeper_membership;
+        self
+    }
+
+    /// Show long strings (defaults to false).
+    pub fn show_long_strings(&mut self, show_long_strings: bool) -> &mut Self {
+        self.long_string_formatter.show_long_strings = show_long_strings;
+        self
+    }
+}
+
+impl fmt::Display for CollectionDisplay<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        display_header(self.collection, f)?;
+        let nerrors = display_errors(self.collection, f)?;
+        display_devices(
+            &self.collection,
+            &self.include_sps,
+            &self.long_string_formatter,
+            f,
+        )?;
+        if self.include_sleds {
+            display_sleds(&self.collection, f)?;
+        }
+        if self.include_orphaned_datasets {
+            display_orphaned_datasets(&self.collection, f)?;
+        }
```

Comment:
I think this is a behavior change we don't want? The old code was:

```
    if filter.include_sleds() {
        inv_collection_print_sleds(&collection);
    } else if matches!(filter, CollectionsShowFilter::OrphanedDatasets) {
        inv_collection_print_orphaned_datasets(&collection);
    }
```

If we print sleds, that already includes the orphaned datasets for each sled. `CollectionsShowFilter::OrphanedDatasets` was a way to say "I _only_ want to see the orphaned datasets from each sled, not all the other details", because we want to be able to do this during the R16 upgrade as a check for any unexpected orphans without all the noise of the normal sled details.

Reply:
Ahh gotcha, will update and explain in a comment.

Reply:
Fixed.